fix(email): Fixes multiple issues related to the failed email retry s…

albertisfu opened the pull request:
This solves multiple issues related to the failed email retry system.

- Avoid creating FailedEmail objects for duplicated bounce events:
    Seems that some email providers can send multiple bounce events for the same message. So in order to avoid creating duplicated `FailedEmail` objects now before creating it we verify that a `FailedEmail` for the same recipient and `message_id` don’t exist already in the DB.

- Solves the duplicated celery tasks issue described in #2233
    - Replaced `send_failed_email` and `check_recipient_deliverability` celery tasks for new periodic methods: `periodic_send_failed_email` and `periodic_check_recipient_deliverability` that are implemented as a new Django command: `cl_retry_failed_email`
  
    - Current celery tasks were not removed from the code base to avoid errors in production once it's released since enqueue tasks will continue to need them. We no longer call these tasks but they will continue being in queue for some days so I only added a DB lock in order to protect them to be executed multiple times.
    - After some days we’ll be able to remove those tasks from the code
    
- Increased `DELIVERABILITY_THRESHOLD`
      Checking the effectivity of the current `DELIVERABILITY_THRESHOLD` (12 hours), based on the current data I noticed that it needed to be increased since we've received some bounce events even 24 hours after the message was sent. So in those cases, the backoff event is restarted even though they belong to the same failure event. So I think 36 hours it’s better so that the backoff continues increasing accordingly.
    
- Added a new setting:`DELIVERABILITY_THRESHOLD`
   Now set to 2 hours, this is the time threshold to consider if a recipient is accepting new messages after a backoff event. If the recipient has not received a new bounce event in this period of time is considered deliverable and WAITING `FailedEmail` objects are scheduled to be sent. So it can be easily tuned if needed.
    
- Updated tests according to the new periodic methods.
    Also fixed tests that use time. Added a new library to fake time in tests: `freezegun` (This library looks more maintained than `faketime`)
    
**Here are some considerations when releasing to production:**
    
- The model `EmailFlag` requires a DB migration of a new field `checked`
    
- After releasing to production (and applying the migration) we need to set via the Django console the checked field of current `EmailFlag` objects to `True` in order to avoid the new `periodic_check_recipient_deliverability` re-schedule messages that are already being processed by celery tasks.

Here is the code to apply:
```
from cl.users.models import EmailFlag
from cl.users.models import FLAG_TYPES
EmailFlag.objects.filter(flag_type=FLAG_TYPES.BACKOFF).update(checked=True)    
```

- After setting the `checked` field to True for current Backoff `EmailFlag` objects we need to set crons for the new Django commands, as follows:

Run every 30 minutes
```
docker exec -it cl-django python /opt/courtlistener/manage.py cl_retry_failed_email --send-failed-email
```
    
Run every 1 hour
```
docker exec -it cl-django python /opt/courtlistener/manage.py cl_retry_failed_email --check-recipients-deliverability
```

`--send-failed-email` needs to run more often since it would send pending FailedEmail including those in ENQUEUED status. That sending them help us to verify if we receive a new bounce event or not so we can know if the recipient is accepting messages again.

`--check-recipients-deliverability` Since the current `DELIVERABILITY_THRESHOLD` is set to 2 hours, it's reasonable to check every hour to look for recipients that are now deliverable. 

These cron run times might represent a delay from the old celery task approach. But at the same time would help to send unnecessary messages if the recipient does not bounce messages immediately.


Let me know what you think!
Thanks.

mlissner commented:
OK, I don't see anything obvious amiss here, but a few comments:

 - If you agree time-machine is better, lets use it.
 - There used to be a delay of 1 minute between each failed email that we'd send to folks. I guess that's gone now?
 - If it's easy to do this as a single cronjob, that's probably easier on the k8s side. Not essential, but is it easy to make it run every 30 minutes and do the needed steps if needed?
 - It sort of feels to me like the `checked` variable should be a time field instead of a boolean. Did you get that sense too? I haven't read the code carefully enough to have a strong feeling, but it feels like that might simplify some things?

albertisfu commented:
Thank you, some comments about it:

> If you agree time-machine is better, lets use it.

Yeah of course!

> There used to be a delay of 1 minute between each failed email that we'd send to folks. I guess that's gone now?

Yes, I removed that delay of one minute because since we're going to send failed messages every 30 minutes, it doesn't make sense a delay of one minute between each email. But this made me think about the SES rate limit, which now is 30 messages per second, django-ses currently throttle emails sent to 50% of the rate limit, so we don't have to apply an additional throttle.
In case we don't want to send emails too fast directly within the cron function we could apply an additional throttle or schedule a celery task for each email delaying the email sent just a few seconds for every message (It wouldn't mean a problem with the visibility timeout since it would be a short time in the future, I think.) 

What do you think?


> If it's easy to do this as a single cronjob, that's probably easier on the k8s side. Not essential, but is it easy to make it run every 30 minutes and do the needed steps if needed?

Yeah, I think it's not a problem to add both functions in the same cron. I'll work on it.

> It sort of feels to me like the checked variable should be a time field instead of a boolean. Did you get that sense too? I haven't read the code carefully enough to have a strong feeling, but it feels like that might simplify some things?

Well, I thought that the boolean field was the simplest way to do it. But yes a time field would have an advantage in case we need to debug things. I'll check this and see if using the time field simplifies things if so I'll change it.

mlissner commented:
> we could apply an additional throttle or schedule a celery task for each email 

If we don't do this, what happens? Will we notice? I'd suggest not doing this until we know we need to, if it'll fail loudly enough.

mlissner commented:
Also, let's do the PR to remove the old code now, and we can just merge it in a week. It'll be ready to go that way.

albertisfu commented:
> If we don't do this, what happens? Will we notice? I'd suggest not doing this until we know we need to, if it'll fail loudly enough.

I expect that there won't be a problem due to the `django-ses` throttling but I'll check if in case the rate limit is exceeded it fails loudly.

> Also, let's do the PR to remove the old code now, and we can just merge it in a week. It'll be ready to go that way.

are you referring to the current retry email celery tasks? if so maybe I'm not understanding what would be the plan here, sorry.

I think we first need to stop enqueuing tasks related to the retry email system (that is in this PR), then mark all the current backoff EmailFlag objects as "checked" so recipients' messages are not rescheduled by the new periodic methods.
So, celery will continue executing retry tasks that are already in the queue for some days. But at the same time, new email failures will be retried by the new periodic methods. From my understanding, they need to live together for some days if we don't want to stop retrying messages

mlissner commented:
> are you referring to the current retry email celery tasks? if so maybe I'm not understanding what would be the plan here, sorry.

Yes. Let's do the PR now, get it approved, and then we'll merge it in five days or so. But the idea is that we want the code ready so we don't forget to do it.

albertisfu commented:
Got it! A different PR. Sorry I was understating to remove those tasks here, yes of course.

albertisfu commented:
Thanks! here are the requested changes. 

- Replaced `freezegun` with `time-machine` to mock time in tests.
- Merged the check_recipient_deliverability and send_failed_email methods in a single command. So it can run every 30 minutes.
- Changed the new EmailFlag checked field from bool to datetime.

I also did some tests to check if we get an error in case we exceed the SES rate limit. I couldn't exceed the rate limit even though I deactivate the django-ses throttling and using a sandbox account with a rate limit of 1 email per second. Seems that the real limit might be a bit higher and sending emails is not that fast. But I could confirm that the default `django-ses` rate limit works well and we would be safe since `django-ses` queries the AWS to get the account rate limit and don't exceed it.

But I was able to reach the daily limit (200 daily emails in sandbox mode) and when it happens an error is triggered:
`botocore.exceptions.ClientError: An error occurred (Throttling) when calling the SendRawEmail operation (reached max retries: 4): Daily message quota exceeded.`

So I assume that there might be a similar error for exceeding the rate limit, so it'll fail loudly.

The code to run after it's released to prod changed a bit:
```
from django.utils.timezone import now
from cl.users.models import EmailFlag
from cl.users.models import FLAG_TYPES
EmailFlag.objects.filter(flag_type=FLAG_TYPES.BACKOFF).update(checked=now())  
```
After it the new command to set up the cron every 30 minutes is the following:
`manage.py cl_retry_failed_email --check-recipients-deliverability-and-send-failed-email`


I've created the PR #2340 that removes the current related celery tasks so that it can be merged in one week or so after this is released.


Let me know if there is anything else.

mlissner commented on cl/users/management/commands/cl_retry_failed_email.py:
```diff
@@ -0,0 +1,103 @@
+
+    help = "Check email recipients' deliverability or send failed emails."
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "--check-recipients-deliverability-and-send-failed-email",
+            action="store_true",
+            default=False,
+            help="Send failed email.",
+        )
```
I think we may as well do this without arguments if this is the only one. 

I think we should make the function name simpler too. Maybe something like, `handle_failing_emails` or something like that?

albertisfu replied:
Changed!

mlissner reviewed (changes requested):
Seems fine. One aesthetic comment for you, please.

albertisfu commented:
Yeah, I removed the arguments from the command and renamed the function to `handle_failing_emails`

The command is just:
`/manage.py cl_retry_failed_email`

Let me know if there is anything else.

Thanks!

mlissner commented:
Looks good. One more complete!
